### PR TITLE
fix(http): allow using jsonpInjectables and httpInjectables in same i…

### LINKED ISSUE #3365: fix(http): cannot use JSONP_BINDINGS and HTTP_BINDINGS in same injector
If I create an injector as follows, Http will use the JsonpBackend instead of XHRBackend.

```js
var injector = Injector.resolveAndCreate([HTTP_BINDINGS, JSONP_BINDINGS]);
var http = injector.get(Http);
http.get('foo.json'); // Would use JSONP Backend.
```

This is because both injectables bind the abstract `ConnectionBackend` class to their own specific implementation.

This should be able to be easily fixed by binding `Jsonp` and `Http` to factories that have dependencies on concrete backends instead of overriding the `ConnectionBackend` bindings.

### PR DESCRIPTION
**Pull request by 0x-r4bbit**
…njector

Fixes #3365

### CODE REVIEW

**Comment by jeffbcross** [comment]
Nice!

File: modules/angular2/http.ts [hunk 1]
```diff
@@ -62,19 +62,21 @@ export {URLSearchParams} from 'angular2/src/http/url_search_params';
  *
  */
```
Comment (jeffbcross) [inline modules/angular2/http.ts]: Should remove the ConnectionBackend import

Comment (PascalPrecht) [inline modules/angular2/http.ts]: @jeffbcross done!

**Comment by jeffbcross** [comment]
I'm going to pull down locally and investigate breakages from the previous Travis run. I think they were just in E2E tests (plus a dartanalyzer failure that will be fixed since you removed the unused ConnectionBackend import).

Could you add a unit test with an injector that is created with both sets of injectables, and then verify that both backend.createConnections are called as expected?

**Comment by PascalPrecht** [comment]
@jeffbcross ha! in fact I started with writing tests and then wasn't sure how to test the backend connection. But yea, I'll try it again :)

**Comment by jeffbcross** [comment]
@PascalPrecht probably easiest to create a parent injector binding the backends to mocks, and then a child injector that just contains the two sets of injectables.

**Comment by jeffbcross** [comment]
Adding XHRBackend and JSONPbackend respectively inside the injectables lists will fix the E2E issue

**Comment by PascalPrecht** [comment]
Okay, I'll have a look at this.

**Comment by PascalPrecht** [comment]
I'm back on this now, hold your horses

**Comment by PascalPrecht** [comment]
@jeffbcross When binding both backends to a mock backend in a parent injector, whereas the child injector still binds the http and jsonp injectables (which have bindings for XHRBackend and JSONPBackend too), don' they shadow the mock bindings from the parent injector?

**Comment by PascalPrecht** [comment]
I think I also need some help on how to spy on a backends `createConnection` method.  Both backends are bound to `MockBackend` but that one doesn't use `SpyObject`. Any hint?

**Comment by jeffbcross** [comment]
> When binding both backends to a mock backend in a parent injector, whereas the child injector still binds the http and jsonp injectables (which have bindings for XHRBackend and JSONPBackend too), don' they shadow the mock bindings from the parent injector?

If you use injectables in a child injector, it would override bindings in the parent injector since it includes bindings for the backends.

Maybe we could pair/vc sometime tomorrow to work this out?

**Comment by PascalPrecht** [comment]
Correct, that's why I was asking. Because Then it doesn't really make sense to have a parent injector that binds to mock backends, if there's a child injector that overrides those anyway.

Sure, let's pair on that tomorrow. Ping me on Hangouts (or even Slack?)

**Comment by PascalPrecht** [comment]
There you go. Thank you so much @jeffbcross !

**Comment by PascalPrecht** [comment]
Updated PR to use clang. This should make Travis happy.

**Comment by jeffbcross** [comment]
Merging, thanks!

**Comment by mhevery** [comment]
@jeffbcross Is this merged? Can you check?

**Comment by jeffbcross** [comment]
@mhevery it didn't merge because a conflict got merged ahead of it. I fixed the conflict and re-pushed to presubmit branch. Thanks for pointing it out!

**Pull request closed without merging**